pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -115,13 +116,28 @@ func (t ResolvedPipelineRunTask) isFailure() bool {
 	}
 	var c *apis.Condition
 	var isDone bool
-	if t.IsCustomTask() {
+
+	switch {
+	case t.IsCustomTask():
 		if t.Run == nil {
 			return false
 		}
 		c = t.Run.Status.GetCondition(apis.ConditionSucceeded)
 		isDone = t.Run.IsDone()
-	} else {
+	case t.IsMatrixed():
+		if len(t.TaskRuns) == 0 {
+			return false
+		}
+		// is failed on the first failure with no remaining retries to fail fast
```

Comment:
could you please update the function docstring to state that a matrixed pipelinetask is considered failed if any of the pipelineTasks have failed? The comment here is also a bit confusing, could you reword?

Lastly, one of the non goals for this TEP is "Terminating early when one of the TaskRuns or Runs created in parallel fails". It seems like if one of the matrixed Tasks fails, the whole PipelineTask will be considered failed in this implementation (fail fast as you said). I'm concerned that other parts of the codebase might be using this function to decide (for example) when a finally task can be scheduled; is it a concern if the matrixed TaskRuns that haven't failed continue to run? Or will there be some sort of cancellation?

Reply:
> could you please update the function docstring to state that a matrixed pipelinetask is considered failed if any of the pipelineTasks have failed? The comment here is also a bit confusing, could you reword?

good catch, will update the docstring and comments 

> Lastly, one of the non goals for this TEP is "Terminating early when one of the TaskRuns or Runs created in parallel fails". It seems like if one of the matrixed Tasks fails, the whole PipelineTask will be considered failed in this implementation (fail fast as you said). I'm concerned that other parts of the codebase might be using this function to decide (for example) when a finally task can be scheduled; is it a concern if the matrixed TaskRuns that haven't failed continue to run? Or will there be some sort of cancellation?

that's right, was planning to add cancel the other taskruns from the matrix upon the first failure - it seems that we have to address the failure strategy in TEP-0090 so will update the TEP shortly with the "fail fast" detail

[matrix in github actions](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#handling-failures) supports failing fast as the default and there are other strategies - will be proposing that we fail fast then explore other failure strategies later if needed